pytorch_lightning/plugins/training_type/fully_sharded.py

```diff
@@ -0,0 +1,212 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import contextlib
+from typing import Generator, List, Optional
+
+import torch
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.plugins.training_type.ddp import DDPPlugin
+from pytorch_lightning.utilities import _FAIRSCALE_FULLY_SHARDED_AVAILABLE
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if _FAIRSCALE_FULLY_SHARDED_AVAILABLE:
+    from fairscale.nn import auto_wrap, default_auto_wrap_policy, enable_wrap, wrap
+    from fairscale.nn.data_parallel import FullyShardedDataParallel
+
+    from pytorch_lightning.overrides.fairscale import LightningFullyShardedModule, unwrap_lightning_module_fully_sharded
+
+
+class FullyShardedPlugin(DDPPlugin):
+
+    def __init__(
+        self,
+        cpu_offload: bool = False,
+        flatten_parameters: bool = True,
+        reshard_after_forward: bool = True,
+        move_grads_to_cpu: Optional[bool] = None,
+        fp32_reduce_scatter: Optional[bool] = None,
+        compute_dtype: Optional[torch.dtype] = None,
+        bucket_cap_mb: int = 25,
+        automatic_module_wrap: bool = False,
+        min_num_params: int = 1e8,
+        parallel_devices: Optional[List[torch.device]] = None,
+        num_nodes: int = 1,
+        cluster_environment: ClusterEnvironment = None,
+        sync_batchnorm: Optional[bool] = False
```

Comment:
@shuyingsunshine21 is working on getting rid of these in https://github.com/PyTorchLightning/pytorch-lightning/pull/7026/

Reply:
> > > Here is a minimal reproducible example:
> > 
> > 
> > Sounds like we may want to relax some constrains in FSDP. Will need to take a closer look when I can find time.
> 
> @SeanNaren , @min-xu-ai , looks like modifying this piece https://github.com/facebookresearch/fairscale/blob/b0e6b9bdeb4270f600ef242b1b1f97f4a61e0a42/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L850-L859
> 
> to
> 
> ```
>   assert m._is_root is None or m._is_root == False
>   if m._is_root is None:
>       m._is_root = False
>       if not self._has_params:
>           assert m._queue_wait_for_post_backward_closure is None
>           m._queue_wait_for_post_backward_closure = self._queue_wait_for_post_backward
> ```
> 
> the above minimal example passed

Thanks for checking this @shuyingsunshine21!!

Thanks for digging into this and finding this fix, I think if the fairscale team are fine with it, we can change this assertion. It makes sense since the model has already been nested with FSDP wrappers and initialized, the question is if it will run fine if we wrap it again (buffers are already flattened, but I think that shouldn't matter)

Reply:
I'll reflect the API changes made in this PR!